python/dgl/nn/pytorch/gt/egt.py

```diff
@@ -0,0 +1,178 @@
+        Edge embedding dimension.
+    num_heads : int
+        Number of attention heads, by which :attr: `ndim` is divisible.
+    num_vns : int
+        Number of virtual nodes.
+    dropout : float, optional
+        Dropout probability. Default: 0.0.
+    attn_dropout : float, optional
+        Attention dropout probability. Default: 0.0.
+    activation : callable activation layer, optional
+        Activation function. Default: nn.ELU().
+    ffn_multiplier : float, optional
+        Multiplier of the inner dimension in Feed Forward Network.
+        Default: 2.0.
+    edge_update : bool, optional
+        Whether to update the edge embedding. Default: True.
+
+    Examples
+    --------
+    >>> import torch as th
+    >>> from dgl.nn import EGTLayer
+
+    >>> batch_size = 16
+    >>> num_nodes = 100
+    >>> ndim, edim = 128, 32
+    >>> nfeat = th.rand(batch_size, num_nodes, ndim)
+    >>> efeat = th.rand(batch_size, num_nodes, num_nodes, edim)
+    >>> net = EGTLayer(
+            ndim=ndim,
+            edim=edim,
+            num_heads=8,
+            num_vns=4,
+        )
+    >>> out = net(nfeat, efeat)
+    """
+
+    def __init__(
+        self,
+        ndim,
+        edim,
+        num_heads,
+        num_vns,
+        dropout=0,
+        attn_dropout=0,
+        activation=nn.ELU(),
+        ffn_multiplier=2.0,
+        edge_update=True,
+    ):
+        super().__init__()
+        self.num_heads = num_heads
+        self.num_vns = num_vns
+        self.edge_update = edge_update
+
+        assert not (ndim % num_heads)
+        self.dot_dim = ndim // num_heads
+        self.mha_ln_h = nn.LayerNorm(ndim)
+        self.mha_ln_e = nn.LayerNorm(edim)
+        self.E = nn.Linear(edim, num_heads)
+        self.QKV = nn.Linear(ndim, ndim * 3)
+        self.G = nn.Linear(edim, num_heads)
+        self.attn_dropout = nn.Dropout(attn_dropout)
+        self.O_h = nn.Linear(ndim, ndim)
+        self.mha_dropout_h = nn.Dropout(dropout)
+
+        node_inner_dim = round(ndim * ffn_multiplier)
+        self.node_ffn = nn.Sequential(
+            nn.LayerNorm(ndim),
+            nn.Linear(ndim, node_inner_dim),
+            activation,
+            nn.Linear(node_inner_dim, ndim),
+            nn.Dropout(dropout),
+        )
+
+        if self.edge_update:
+            self.O_e = nn.Linear(num_heads, edim)
+            self.mha_dropout_e = nn.Dropout(dropout)
+            edge_inner_dim = round(edim * ffn_multiplier)
+            self.edge_ffn = nn.Sequential(
+                nn.LayerNorm(edim),
+                nn.Linear(edim, edge_inner_dim),
+                activation,
+                nn.Linear(edge_inner_dim, edim),
+                nn.Dropout(dropout),
+            )
+
+    def forward(self, h, e, mask=None):
+        """Forward computation. Note: :attr:`h` and :attr:`e` should be padded
+        with embedding of virtual nodes if :attr:`num_vns` > 0, while
+        :attr:`mask` should be padded with `0` values for virtual nodes.
```

Comment:
the padding should be in the beginning or at the end?

Reply:
Actually it doesn't matter since users can define which nodes to be used in the readout function in EGT model. The official implementation of EGT is to put it at the beginning.

Reply:
It seems that the `degree_scalars` of virtual nodes is set to 1.0 at the beginning. Does it suggest that padding at the beginning is mandatory?

Reply:
btw, do you think it's a better solution to include a VirtualNode module like what EGT does? Will it reduce user's burden and improve usability?

Reply:
> It seems that the `degree_scalars` of virtual nodes is set to 1.0 at the beginning. Does it suggest that padding at the beginning is mandatory?

Yes, I was wrong. It should be in the beginning.

Reply:
> btw, do you think it's a better solution to include a VirtualNode module like what EGT does? Will it reduce user's burden and improve usability?

I think so, but this VirtualNode module may not be universal since its function is not only to generate virtual node embedding, but also to general positional encodings or attention bias related to the virtual nodes. While PEs and attention bias are different in different models.